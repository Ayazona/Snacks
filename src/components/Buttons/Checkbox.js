import React from 'react'
import PropTypes from 'prop-types'
import RadioCheckboxBase from '../../base/RadioCheckboxBase'
import bkgSvgBase from '../../assets/checkboxBase.svg'
import bkgSvgSelected from '../../assets/checkboxSelected.svg'

const renderInputButton = (isSelected, style) => {
  const SvgComponent = isSelected ? bkgSvgSelected : bkgSvgBase
  return <SvgComponent aria-hidden="true" style={style}/>
}

const Checkbox = (props) => {
  return (
    <RadioCheckboxBase
      btnType='checkbox'
      renderInputButton={renderInputButton}
      {...props}
    />
  )
}

Checkbox.propTypes = {
  aria          : PropTypes.shape({
    label         :PropTypes.string,
  }),
  children      : PropTypes.string,
  id            : PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  isDisabled    : PropTypes.bool,
  isSelected    : PropTypes.bool,
  name          : PropTypes.string,
  onChange      : PropTypes.func,
  onBlur        : PropTypes.func,
  onFocus       : PropTypes.func,
  style         : PropTypes.shape({
    button        : PropTypes.object,
    label         : PropTypes.object,
    wrapEl        : PropTypes.object,
  }),
  value         : PropTypes.string,
  wrapEl        : PropTypes.string,
}

export default Checkbox
