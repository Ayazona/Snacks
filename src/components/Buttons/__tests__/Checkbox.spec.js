import Checkbox from '../Checkbox'
import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import sinon from 'sinon'

describe('Checkbox', () => {
  it('renders the correct button type', () => {
    const tree = renderer
      .create(<Checkbox id={1} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders the correct selected state', () => {
    const testCases = [
      { id: 1, isSelected: false },
      { id: 2, isSelected: true },
      { id: 2, isSelected: true, isDisabled: true },
    ]

    testCases.forEach(props => {
      const tree = renderer
        .create(<Checkbox {...props} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  it('disables the input when it is disabled', () => {
    const wrapper = mount(<Checkbox id="A" value="test" isSelected>Testing disabled</Checkbox>)
    const htmlBtn = wrapper.find('input').instance()

    expect(htmlBtn.checked).toBe(true)
    expect(htmlBtn.disabled).toBe(false)
    
    wrapper.setProps({isDisabled: true})

    expect(htmlBtn.checked).toBe(true)
    expect(htmlBtn.disabled).toBe(true)
  })

  
  it('generates a label if a child prop of text is supplied', () => {
    const tree = renderer
      .create(<Checkbox id={1}>Test label</Checkbox>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  

  it('incorporates user style if passed', () => {
    const testCases = [
      {
        id: 1,
        style: {
          wrapEl: {background: 'pink'},
          button: {border: '2px solid purple' },
          label: {color: 'green'}
        },
      },
      {
        id: 2,
        style: undefined
      },
    ]

    testCases.forEach(props => {
      const tree = renderer
        .create(<Checkbox {...props}>Test label</Checkbox>)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  it('calls the user-supplied onChange function', () => {
    const onChange = sinon.spy()
    const wrapper = mount(<Checkbox id={1} onChange={onChange} />)

    wrapper.find('input').simulate('change')
    expect(onChange.calledOnce).toBe(true)
  })

  it('correctly forwards refs', () => {
    const ref = React.createRef()

    // Enzyme does not currently support forwarded refs. react-test-renderer does, but you must specify createNodeMock.
    // https://github.com/airbnb/enzyme/issues/1852
    renderer.create(<Checkbox id={1} ref={ref} />, { createNodeMock: element => element })

    expect(ref.current.type).toBe('input')
  })
})
