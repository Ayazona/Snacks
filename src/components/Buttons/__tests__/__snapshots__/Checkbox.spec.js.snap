// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox generates a label if a child prop of text is supplied 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "display": "flex",
    }
  }
>
  <div
    data-radium={true}
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <img
      alt=""
      data-radium={true}
      src="test-file-stub"
      style={
        Object {
          "height": 22,
          "width": 22,
        }
      }
      tabIndex="-1"
    />
    <input
      aria-label={undefined}
      checked={false}
      data-radium={true}
      id={1}
      onClick={[Function]}
      style={
        Object {
          "MozAppearance": "none",
          "MsAppearance": "none",
          "WebkitAppearance": "none",
          "appearance": "none",
          "height": 22,
          "left": 0,
          "position": "absolute",
          "top": 0,
          "width": 22,
        }
      }
      type="checkbox"
      value={undefined}
    />
  </div>
  <label
    data-radium={true}
    htmlFor={1}
    style={
      Object {
        "marginLeft": 10,
      }
    }
  >
    Test label
  </label>
</div>
`;

exports[`Checkbox renders the correct button type 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <img
    alt=""
    data-radium={true}
    src="test-file-stub"
    style={
      Object {
        "height": 22,
        "width": 22,
      }
    }
    tabIndex="-1"
  />
  <input
    aria-label={undefined}
    checked={false}
    data-radium={true}
    id={1}
    onClick={[Function]}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "height": 22,
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": 22,
      }
    }
    type="checkbox"
    value={undefined}
  />
</div>
`;

exports[`Checkbox renders the correct selected state 1`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <img
    alt=""
    data-radium={true}
    src="test-file-stub"
    style={
      Object {
        "height": 22,
        "width": 22,
      }
    }
    tabIndex="-1"
  />
  <input
    aria-label={undefined}
    checked={false}
    data-radium={true}
    id={1}
    onClick={[Function]}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "height": 22,
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": 22,
      }
    }
    type="checkbox"
    value={undefined}
  />
</div>
`;

exports[`Checkbox renders the correct selected state 2`] = `
<div
  data-radium={true}
  style={
    Object {
      "position": "relative",
    }
  }
>
  <img
    alt=""
    data-radium={true}
    src="test-file-stub"
    style={
      Object {
        "height": 22,
        "width": 22,
      }
    }
    tabIndex="-1"
  />
  <input
    aria-label={undefined}
    checked={true}
    data-radium={true}
    id={2}
    onClick={[Function]}
    style={
      Object {
        "MozAppearance": "none",
        "MsAppearance": "none",
        "WebkitAppearance": "none",
        "appearance": "none",
        "height": 22,
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": 22,
      }
    }
    type="checkbox"
    value={undefined}
  />
</div>
`;
